### PR TITLE
Adding support for downloading media from existing CVAT tasks/projects

### PR DESCRIPTION
**Pull request by ehofesmann**
Adds a `download_media=True` option to `import_annotations()`  that downloads the media associated with a given CVAT project or tasks along with the annotations.

### Note

[CVAT force re-encodes all videos as MP4s at 25fps](https://github.com/openvinotoolkit/cvat/blob/cde33acf5da2c29b5c70dec7df688af60734e1a7/cvat/apps/engine/media_extractors.py#L590-L595), so, in general, this workflow is not an advisable way to distribute video datasets.

### Image dataset example

```py
import fiftyone as fo
import fiftyone.zoo as foz
import fiftyone.utils.cvat as fouc

dataset = foz.load_zoo_dataset("quickstart")

# Create a CVAT project with labels
results = dataset.annotate(
    "test",
    label_field="ground_truth",
    segment_size=3,
    project_name="test",
)

# Download dataset from CVAT
dataset = fo.Dataset()
fouc.import_annotations(
    dataset,
    project_name="test",
    data_path="/tmp/cvat_example",
    download_media=True,
)

session = fo.launch_app(dataset)
```

### Video dataset example

```py
import fiftyone as fo
import fiftyone.zoo as foz
import fiftyone.utils.cvat as fouc

dataset = foz.load_zoo_dataset("quickstart-video", max_samples=3)

# Create a CVAT project with labels
results = dataset.annotate(
    "test",
    label_field="frames.detections",
    project_name="test-video",
)

# Download dataset from CVAT
dataset = fo.Dataset()
fouc.import_annotations(
    dataset,
    project_name="test-video",
    data_path="/tmp/cvat_video_example",
    download_media=True,
)

session = fo.launch_app(dataset)
```

### CODE REVIEW

File: fiftyone/utils/cvat.py [hunk 1]
```diff
@@ -5773,6 +5862,7 @@ def _build_frame_id_map(self, samples):
         frame_id = -1
 
         if samples.media_type == fom.VIDEO:
+            samples.ensure_frames()
```
Comment (brimoor) [inline fiftyone/utils/cvat.py:5865]: This is no longer necessary after my last updates in https://github.com/voxel51/fiftyone/pull/1510.

**Review by brimoor — CHANGES_REQUESTED** [review]
Looks like there are a couple merge conflicts to resolve after my last changes in https://github.com/voxel51/fiftyone/pull/1510

File: fiftyone/utils/cvat.py [hunk 2]
```diff
@@ -3236,6 +3315,16 @@ def task_url(self, task_id):
     def task_data_url(self, task_id):
         return "%s/data" % self.task_url(task_id)
 
+    def task_data_download_url(
+        self, task_id, frame_id, data_type="frame", quality="original"
+    ):
+        return "%s/data?type=%s&quality=%s&number=%d" % (
```
Comment (brimoor) [inline fiftyone/utils/cvat.py:3321]: Just checking: are `type` and `number` mandatory for videos here?

I guess you wouldn't have gone through the trouble of merging chunks if you could have avoided it, but if CVAT stores the original video that'd be much cleaner

File: fiftyone/utils/cvat.py [hunk 3]
```diff
@@ -222,21 +240,82 @@ def import_annotations(
...
+):
     resp = api.get(api.task_data_meta_url(task_id)).json()
+    start_frame = resp.get("start_frame", None)
+    stop_frame = resp.get("stop_frame", None)
+    chunk_size = resp.get("chunk_size", None)
     filenames = [frame["name"] for frame in resp["frames"]]
 
     filepaths = []
-    for filename in filenames:
+    for frame_id, filename in enumerate(filenames):
```
Comment (brimoor) [inline fiftyone/utils/cvat.py:258]: Collecting the download tasks and running them via a thread pool would help with efficiency here:

```py
def _download_media(tasks, num_workers):
    logger.info("Downloading media...")
    if not num_workers or num_workers <= 1:
        with fou.ProgressBar() as pb:
            for task in pb(tasks):
                _do_download_media(task)
    else:
        with multiprocessing.dummy.Pool(processes=num_workers) as pool:
            with fou.ProgressBar(total=len(tasks)) as pb:
                results = pool.imap_unordered(_do_download_media, tasks)
                for _ in pb(results):
                    pass
```

File: fiftyone/utils/cvat.py (hunk 2 above)
Comment (ehofesmann) [reply fiftyone/utils/cvat.py:3321]: Right, `type` `quality` and `number` are mandatory and can only give you either the compressed or original video `chunks`, not the original video itself, so you need to provide the `number` of the chunk you want to download.

File: docs/source/integrations/cvat.rst [hunk 4]
```diff
@@ -2144,19 +2144,34 @@ to import individual task(s) or an entire project into a FiftyOne dataset.
...
 
     #
-    # Create a mapping between filenames in the pre-existing CVAT project and
-    # the locations of the media locally on disk for the FiftyOne dataset
+    # In the simplest case, you can download both the annotations and the media
+    # from CVAT
+    #
+
+    dataset = fo.Dataset()
+    fouc.import_annotations(
```
Comment (brimoor) [inline docs/source/integrations/cvat.rst:2152]: Much cleaner than the `data_map` variation for docs purposes 🥳

File: fiftyone/utils/cvat.py [hunk 5]
```diff
@@ -247,6 +310,57 @@ def _parse_task_metadata(
...
+        with etau.TempDir() as tmp_dir:
+            chunk_paths = []
+            for chunk_id in range(num_chunks):
+                resp = api.get(
+                    api.task_data_download_url(
+                        task_id, chunk_id, data_type="chunk"
+                    )
+                )
+                chunk_path = os.path.join(tmp_dir, "%d.%s" % (chunk_id, ext))
+                etau.write_file(resp._content, chunk_path)
```
Comment (brimoor) [inline fiftyone/utils/cvat.py:355]: As a general rule, FiftyOne is written to automatically create output directories if they don't already exist. Using `etau.write_file()` is an idiomatic way to handle that.

File: fiftyone/utils/video.py [hunk 6]
```diff
@@ -519,6 +519,26 @@ def sample_frames_uniform(
     return sample_frames
 
 
+def concat_videos(input_paths, output_path, verbose=False):
```
Comment (brimoor) [inline fiftyone/utils/video.py:522]: I think `concat` is a bit more clear here. "Merge" typically means some non-trivial combination of things.

**Review by brimoor — APPROVED** [review]
LGTM. Works like a charm!

**Pull request merged**